src/winkerberos.c

```diff
@@ -437,8 +437,160 @@ sspi_client_clean(PyObject* self, PyObject* args) {
+"  - `initiator_address`: Optional byte string of the initiator address.\n"
+"  - `acceptor_addrtype`: Optional int of the acceptor address type,\n"
+"    default is GSS_C_AF_UNSPEC (0).\n"
+"  - `acceptor_address`: Optional byte string of the acceptor address.\n"
+"  - `application_data`: Optional byte string that should be set under the\n"
+"    application_data field. An example of this would be\n"
+"    'tls-server-endpoint:{cert-hash}' where {cert-hash} is the hash of the\n"
+"    server's certificate.\n"
+"\n"
+":Returns: SecPkgContext_Bindings* where SecPkgContext_Bindings is the\n"
+"channel bindings structure that can be passed onto L{authGSSClientStep}.");
+static PyObject*
+sspi_channel_bindings(PyObject* self, PyObject* args, PyObject* keywds) {
+    static char *kwlist[] = {"initiator_addrtype", "initiator_address", "acceptor_addrtype",
+        "acceptor_address", "application_data", NULL};
+
+    SEC_CHANNEL_BINDINGS* channel_bindings;
+    SecPkgContext_Bindings* context_bindings;
+    PyObject *pychan_bindings = NULL;
+
+    unsigned long initiator_addrtype = 0;
+    unsigned long acceptor_addrtype = 0;
+    unsigned long initiator_length = 0;
+    unsigned long acceptor_length = 0;
+    unsigned long application_length = 0;
+    char *initiator_address = NULL;
+    char *acceptor_address = NULL;
+    char *application_data = NULL;
+
+    size_t offset = 0;
+    size_t data_length = 0;
+    unsigned char *offset_p = NULL;
+
+    if (!PyArg_ParseTupleAndKeywords(args, keywds, "|ls#ls#s#", kwlist,
+            &initiator_addrtype, &initiator_address, &initiator_length,
+            &acceptor_addrtype, &acceptor_address, &acceptor_length,
+            &application_data, &application_length)) {
+        return NULL;
+    }
+
+    data_length = initiator_length + acceptor_length + application_length;
+    offset = sizeof(SEC_CHANNEL_BINDINGS);
+
+    context_bindings = (SecPkgContext_Bindings *) malloc(sizeof(SecPkgContext_Bindings));
+    if (context_bindings == NULL) {
+        PyErr_SetString(PyExc_MemoryError, "Failed to allocate a memory block for SecPkgContext_Bindings");
+        return NULL;
+    }
+    context_bindings->BindingsLength = sizeof(SEC_CHANNEL_BINDINGS) + data_length;
+
+    channel_bindings = (SEC_CHANNEL_BINDINGS*) malloc(context_bindings->BindingsLength);
+    if (channel_bindings == NULL) {
+        free(context_bindings);
+        PyErr_SetString(PyExc_MemoryError, "Failed to allocate a memory block for SEC_CHANNEL_BINDINGS");
+        return NULL;
+    }
+    context_bindings->Bindings = channel_bindings;
+
+    channel_bindings->dwInitiatorAddrType = initiator_addrtype;
+    channel_bindings->cbInitiatorLength = initiator_length;
+    if (initiator_address != NULL) {
+        channel_bindings->dwInitiatorOffset = offset;
+
+        offset_p = &((unsigned char*) channel_bindings)[offset];
+        memcpy_s(&offset_p[0], initiator_length, initiator_address, initiator_length);
```

Comment:
The idea behind memcpy_s is to ensure you have enough buffer space to copy the data on each call. Instead of passing initiator_length as the second param you want to pass data_length. Then, in the next call, data_length - initiator_length, then in the next call data_length - initiate_length - acceptor_length, etc. The function can only protect you if you tell it how much buffer space is actually left.

See https://msdn.microsoft.com/en-us/library/wes2t00f.aspx